Additional tests, and some fixes, related to AlwaysOn

johlju opened the pull request:
* Added tests for resources
  - xSQLServerPermission
  - xSQLServerEndpointState
  - xSQLServerEndpointPermission
  - xSQLServerAvailabilityGroupListener
* Fixes in xSQLServerAvailabilityGroupListener
  - In one case the Get-method did not report that DHCP was configured. 
  - Now the resource will throw 'Not supported' when IP is changed between Static and DHCP.
  - Fixed an issue where sometimes the listener wasn't removed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xsqlserver/98)
<!-- Reviewable:end -->

msftclas commented:
Hi __@johlju__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

narrieta commented:
What is the test that fails without a fake function?

johlju commented:
See the block `Describe "$($script:DSCResourceName)\Set-TargetResource"` in the file xSQLServerEndpointState.Tests.ps1

johlju commented:
If I remove this block 
```
        Get-Module -Name MockSQLPS | Remove-Module -Force
        New-Module -Name MockSQLPS -ScriptBlock {
            function Set-SqlHADREndpoint { return }
        
            Export-ModuleMember -Function Set-SqlHADREndpoint
        } | Import-Module -Force
```
I get this error
```
Describing xSQLServerEndPointState\Set-TargetResource
 [-] Error occurred in Describe block 365ms
   CommandNotFoundException: Could not find Command Set-SqlHADREndpoint
   at Validate-Command, C:\Program Files\WindowsPowerShell\Modules\Pester\3.4.0\Functions\Mock.ps1: line 801
   at Mock, C:\Program Files\WindowsPowerShell\Modules\Pester\3.4.0\Functions\Mock.ps1: line 168
   at <ScriptBlock>, V:\Source\xSqlServer\Tests\Unit\xSQLServerEndpointState.Tests.ps1: line 262
```

kwirkykat commented:
@johlju @aultt is having issues with these same tests in his PR #89 . Have you seen these test issues before? Will this PR fix those issues?

narrieta commented:
@johlju OK, so I followed Pester's code and you are correct, the command cannot be resolved (since the SqlServer module is not installed when the UTs run).  Your solution will make things work, but I think there should be a comment explaining why the mock module is there.

Some suggestions you may want to consider: 
- I would define a psm1 file for the fake SqlServer instead of defining it within the tests. This would allow sharing the fake module across multiple tests.
- I would throw from the fake cmdlets instead of making them no-ops. Those cmdlets should never be reached by any tests so having an error would be good.

    function Set-SqlHADREndpoint { throw 'NotImplemented'  }

BrianFarnhill commented:
@narrieta That's actually genius - we do the PSM1 file for SharePointDsc but we don't throw inside the calls, we just use them to create our mocks from. I might need to alter my script which generates those stubs to add the exception call in.

johlju commented:
@kwirkykat No, those issues in the test run for PR #89 got to do with an issue with the testing framework, see @mbreakey3 and our comments about the problem in issue #92.

Maybe when this PR is merged, and then @aultt resolve conflicts for #89 it might auto heal?

johlju commented:
@narrieta I will change so that it used a stub module instead, and that the functions in the stub throws, as per your suggestion.

aultt commented:
@johlju looks like you have a pending review can you acknowledge so this can be merged?  Waiting on your changes so my PR can go.

johlju commented:
@aultt I'm working on the code change which is 90% done, I will try to commit it to the PR tomorrow night (Sweden time). Otherwise on Friday evening. (I'm currently away in another city so have limited time to code :/ )

aultt commented:
@johlju sounds good Ill take a look next week.

johlju commented:
@kwirkykat or/and @mbreakey3 Can you tell med how I can suppress the PSScriptAnalyzer check for Username and Password parameters in the Stub file? See the AppVeyor failed checks.
These stub files are generated from the actual modules, so the real cmdlets actually uses `[String]`. 
So these needs to be suppressed so the functions can mimic the real cmdlets. Or do they? Maybe there is another solution I can't think of?

_You can read about how they are generated in the issue disussion #90_.

johlju commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[Tests/Unit/Stubs/SQLServerStub.psm1, line 0 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzYx6C0I6AaXlBLu2n:-KOzYx6C0I6AaXlBLu2o:b-vp2ouj) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/Stubs/SQLServerStub.psm1#L0)):*
This stub file is not actual used by any test yet, but I figured I add this because of issue #91 will make sure that SQLServer module will used before SQLPS, if it exist. So we should actually run tests with both SQLPS and SQLServer stub module.
As of now, these tests I added only run the test using SQLPS stub module.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
# Suppressing this rule because these functions are from an external module 
# and are only being used as stubs
[System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingUserNameAndPassWordParams', '')]
param()

mbreakey3 commented:
Reviewed 1 of 8 files at r1.
Review status: 1 of 10 files reviewed at latest revision, 12 unresolved discussions.

---

*[README.md, line 345 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOznOcW6435bvsxIJbH:-KOznOcW6435bvsxIJbI:ba7efvq) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/README.md#L345)):*
> ```Markdown
>     - modified functions
>         - New-TerminatingError - *added optional parameter `InnerException` to be able to give the user more information in the returned message*
> * Added tests for resources
> ```

Move these comments under 'Unreleased' - I think it got messed up when we released on Wed.

---

*[DSCResources/xSQLServerAvailabilityGroupListener/xSQLServerAvailabilityGroupListener.psm1, line 288 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOznmOmqbLmP6XFmtl7:-KOznmOmqbLmP6XFmtl8:beatu0p) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/DSCResources/xSQLServerAvailabilityGroupListener/xSQLServerAvailabilityGroupListener.psm1#L288)):*
> ```PowerShell
> 
>         if( -not $($PSBoundParameters.ContainsKey('Ensure')) -or $Ensure -eq "Present" ) { 
>             if( ($Port -eq "" -or $listenerState.Port -eq $Port) -and $ipAddressEqual -and ( -not $($PSBoundParameters.ContainsKey('DHCP')) -or $listenerState.DHCP -eq $DHCP ) ) {
> ```

can you divide up this line so that each 'and' is on its own line?

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 50 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzo78HNeCVho3V6SqN:-KOzo78HNeCVho3V6SqO:bpkos05) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L50)):*
> ```PowerShell
>     Describe "$($script:DSCResourceName)\Get-TargetResource" {
>         Context 'When the system is not in the desired state' {
>             BeforeAll {
> ```

You can remove this block if you're not using it

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 90 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzoMZ7ujkv_XHjD0P2:-KOzoMZ7ujkv_XHjD0P3:b37iae5) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L90)):*
> ```PowerShell
>         Context 'When the system is in the desired state, without DHCP' {
>             BeforeAll {
>                 # This has intentially been left blank
> ```

Again - this can be removed

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KOzodQ430m2hgsXPgXD:b-6lfx62) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
> ```PowerShell
>                 return New-Object Object | 
>                     Add-Member NoteProperty Name $listnerName -PassThru | 
>                     Add-Member NoteProperty PortNumber 5030 -PassThru | 
> ```

Can you declare each of these hard coded numbers (port, ipAddress, subnetMask) as variables at the top of the test? 

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 143 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzotk4ff4BohyS5BlL:-KOzotk4ff4BohyS5BlM:byp18fd) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L143)):*
> ```PowerShell
> 
>         Context 'When the system is in the desired state, with DHCP' {
>             BeforeAll {
> ```

Remove

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 201 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzp3BJh60InPRWRJTN:-KOzp3BJh60InPRWRJTO:b-l6mwxj) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L201)):*
> ```PowerShell
>     Describe "$($script:DSCResourceName)\Test-TargetResource" {
>         Context 'When the system is not in the desired state' {
>             BeforeAll {
> ```

remove

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 209 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzp6kVY08tMZ0DcFzB:-KOzp6kVY08tMZ0DcFzC:b-toojov) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L209)):*
> ```PowerShell
>                 $testParameters += @{
>                     Ensure = 'Present'
>                     IpAddress = '192.168.10.45/255.255.252.0'
> ```

Create variable for hard-coded values

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 260 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzpOSmdYG5HflQ-kk1:-KOzpOSmdYG5HflQ-kk2:b-vcknem) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L260)):*
> ```PowerShell
>                 }            
> 
>                 Mock -CommandName Get-SQLAlwaysOnAvailabilityGroupListener -MockWith {
> ```

When you're using the same mock in multiple It statements or context blocks you can just declare once at the top of the Describe block

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 292 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzphtIPEZlum9hWj5I:-KOzphtIPEZlum9hWj5J:b-m5sxs3) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L292)):*
> ```PowerShell
>                 }            
> 
>                 Mock -CommandName Get-SQLAlwaysOnAvailabilityGroupListener -MockWith {
> ```

Since you have some of these where IsDHCP is set to false and some where it's set to true, divide those into two different 'Context' blocks and declare the mock at the top of each and then include each 'It' statement that uses that particular Mock in that 'Context' block. Alternatively, you could set a [Bool] $mockDHCP variable at the top of the Describe block and then just change that value for each 'It' rather than declaring the mock over and over.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 469 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzqpVD-PfRGA2uZuiO:-KOzqpVEPievMMS1kXTD:b-l6mwxj) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L469)):*
> ```PowerShell
>         Context 'When the system is in the desired state' {
>             BeforeAll {
>                 # This has intentially been left blank
> ```

remove

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 1 of 10 files reviewed at latest revision, 12 unresolved discussions.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KP-5Y18fPz2gnzyYHCI:bwrlfvb) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Can you declare each of these hard coded numbers (port, ipAddress, subnetMask) as variables at the top of the test? 
</details>
Love to do this. But I have problem doing this... Strange things going on when I add variables for these values. Doesn't work running the test in PowerShell, but it works running the test in ISE. :/
I will update the PR with the change in one place only (to begin with), to see what AppVeyor thinks about it. If it works I update the rest. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 1 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[README.md, line 345 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOznOcW6435bvsxIJbH:-KP-7Wq4D7aygVVCRB6z:bb6un1j) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/README.md#L345)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Move these comments under 'Unreleased' - I think it got messed up when we released on Wed.
</details>
Done

---

*[DSCResources/xSQLServerAvailabilityGroupListener/xSQLServerAvailabilityGroupListener.psm1, line 288 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOznmOmqbLmP6XFmtl7:-KP-7dPt_2X8pRDBqSov:bzc4had) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/DSCResources/xSQLServerAvailabilityGroupListener/xSQLServerAvailabilityGroupListener.psm1#L288)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> can you divide up this line so that each 'and' is on its own line?
</details>
Done (hope it's better)

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 50 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzo78HNeCVho3V6SqN:-KP-7pT7hALYumjNAo3I:bxe9lr9) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L50)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> You can remove this block if you're not using it
</details>
Done. Removed everywhere

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 90 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzoMZ7ujkv_XHjD0P2:-KP-7sQspjlhrdAdbdbp:bb6un1j) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L90)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Again - this can be removed
</details>
Done

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KP-7wDNi2kYnopXSOZZ:b-gsjd3w) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Love to do this. But I have problem doing this... Strange things going on when I add variables for these values. Doesn't work running the test in PowerShell, but it works running the test in ISE. :/
> I will update the PR with the change in one place only (to begin with), to see what AppVeyor thinks about it. If it works I update the rest. 
</details>
@mbreakey3 It failed in AppVeyor like it did when I run it in PowerShell. Can you help me debug why the variables don't work in the Mock when run from PowerShell but ISE? Staring me blind. Probably something simple. 
_The existing variable `$listnerName` don't work either. Never saw that before because the Name property is not used in the code (will remove it from the test in the next commit)._

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 143 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzotk4ff4BohyS5BlL:-KP-B1zB_6zL1t2qi4Si:bb6un1j) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L143)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Remove
</details>
Done

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 201 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzp3BJh60InPRWRJTN:-KP-B2gDso2jyoKvY-Ny:bb6un1j) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L201)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> remove
</details>
Done

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 469 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzqpVD-PfRGA2uZuiO:-KP-B4ODaH4ZpU044Iy8:bb6un1j) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L469)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> remove
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 1 of 10 files reviewed at latest revision, 12 unresolved discussions.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KP-I3ohjSHvsMfVypZZ:b-gc9lmj) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> @mbreakey3 It failed in AppVeyor like it did when I run it in PowerShell. Can you help me debug why the variables don't work in the Mock when run from PowerShell but ISE? Staring me blind. Probably something simple. 
> _The existing variable `$listnerName` don't work either. Never saw that before because the Name property is not used in the code (will remove it from the test in the next commit)._
</details>
Okay, so if I set these as `$global:` it works. AppVeyor likes it too. Is there a better way? I can go with global, but doesn't feel right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Reviewed 2 of 8 files at r4.
Review status: 3 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KP-R_gxAM4e9StK_Vy1:b-iobuo6) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Okay, so if I set these as `$global:` it works. AppVeyor likes it too. Is there a better way? I can go with global, but doesn't feel right.
</details>
See my comment above

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 209 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzp6kVY08tMZ0DcFzB:-KP-Rdez-9pWxkRc-DRR:bb6ywmx) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L209)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Create variable for hard-coded values
</details>
LGTM

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 41 \[r4\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KP-QyMC8AO4S3OPGno7:-KP-QyMC8AO4S3OPGno8:b-6zgutf) ([raw file](https://github.com/powershell/xsqlserver/blob/4b6c96f3fa65059742997cdfa59b2edfe96f569d/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L41)):*
> ```PowerShell
>     
>     # Values used for mocking
>     $global:desiredPortNumber = 5030
> ```

Yeah, so these have to be within the Describe block to work with Pester as local variables. Could you try making them $script: variables? If that doesn't work, you can move them into each Describe block separately.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 3 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KP1kqsbkq0K4_pBx1gd:b-2up7pd) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> See my comment above
</details>
So these variables doesn't want to work for me at all :/ Commited my last test to show the problem.

I tried using `$script:`. It doesn't work.
I tried using them inside the Describe-block. It doesn't work.
I tried using them inside the Context-block. It doesn't work.
I tried using them inside the Mock-block it self. Well that does work for the `$actualPortNumber` variable, but doesn't work on `$actualIPAddress` or `$actualSubnetMask`. _This is the last commit_

But each test does work in ISE (not tested $script thou), just not at the PowerShell prompt or in AppVeyor. I guess ISE are scoping these differently or something.

Any more suggestions, anyone? :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

johlju commented:
Review status: 3 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KP2Bd8s3nuyrrBMzoOz:b-9huikr) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> So these variables doesn't want to work for me at all :/ Commited my last test to show the problem.
> 
> I tried using `$script:`. It doesn't work.
> I tried using them inside the Describe-block. It doesn't work.
> I tried using them inside the Context-block. It doesn't work.
> I tried using them inside the Mock-block it self. Well that does work for the `$actualPortNumber` variable, but doesn't work on `$actualIPAddress` or `$actualSubnetMask`. _This is the last commit_
> 
> But each test does work in ISE (not tested $script thou), just not at the PowerShell prompt or in AppVeyor. I guess ISE are scoping these differently or something.
> 
> Any more suggestions, anyone? :)
</details>
Seems this is a bug (or missing feature). See pester/Pester#204. @mbreakey3 what are your thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

aultt commented:
Congrats @johlju Looks like the build worked!!!

johlju commented:
Review status: 3 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 99 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzodQ430m2hgsXPgXC:-KPE0y6KyxOx7WJyauPS:b55fgxj) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L99)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Seems this is a bug (or missing feature). See pester/Pester#204. @mbreakey3 what are your thoughts?
</details>
Since I had problems with local variables I reverted back to the initial code with hardcoded values in mocks.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 209 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzp6kVY08tMZ0DcFzB:-KPE12pNLCdPvahyNCu8:bg2yt4d) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L209)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> LGTM
</details>
Since I had problems with local variables (see other comments) I reverted back to the initial code with hardcoded values in mocks.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 260 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzpOSmdYG5HflQ-kk1:-KPE15RwOoWwV6ARxKmy:bb6un1j) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L260)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> When you're using the same mock in multiple It statements or context blocks you can just declare once at the top of the Describe block
</details>
Done

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 292 \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzphtIPEZlum9hWj5I:-KPE16_G9yIWxgc6p-Qu:br1fom3) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L292)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Since you have some of these where IsDHCP is set to false and some where it's set to true, divide those into two different 'Context' blocks and declare the mock at the top of each and then include each 'It' statement that uses that particular Mock in that 'Context' block. Alternatively, you could set a [Bool] $mockDHCP variable at the top of the Describe block and then just change that value for each 'It' rather than declaring the mock over and over.
</details>
Done (I hope). Did not change the hardcoded values, see other comments.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 41 \[r4\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KP-QyMC8AO4S3OPGno7:-KPE0d796FrWWVGDMRwe:bg2yt4d) ([raw file](https://github.com/powershell/xsqlserver/blob/4b6c96f3fa65059742997cdfa59b2edfe96f569d/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L41)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Yeah, so these have to be within the Describe block to work with Pester as local variables. Could you try making them $script: variables? If that doesn't work, you can move them into each Describe block separately.
</details>
Since I had problems with local variables (see other comments) I reverted back to the initial code with hardcoded values in mocks.

---

*[Tests/Unit/Stubs/SQLServerStub.psm1, line  \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzYx6C0I6AaXlBLu2n:-KPE1G75jyhlSZSaimxF:b-bg0pdf) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/Stubs/SQLServerStub.psm1#L)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> This stub file is not actual used by any test yet, but I figured I add this because of issue #91 will make sure that SQLServer module will used before SQLPS, if it exist. So we should actually run tests with both SQLPS and SQLServer stub module.
> As of now, these tests I added only run the test using SQLPS stub module.
</details>
Shall I keep this file in?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

johlju commented:
@aultt Yes, it worked first time around too... :) Tried to get local variables to work in mocks, as per suggestion of @mbreakey3. That did not work out to well, that's why it failed. Reverted back to hardcoded values again to get AppVeyor happy. :D
And fixed the rest of the review comments.

mbreakey3 commented:
*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 41 \[r4\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KP-QyMC8AO4S3OPGno7:-KPEwiH2DdGL6q7fpOAR:brslit9) ([raw file](https://github.com/powershell/xsqlserver/blob/4b6c96f3fa65059742997cdfa59b2edfe96f569d/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L41)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Since I had problems with local variables (see other comments) I reverted back to the initial code with hardcoded values in mocks.
</details>
Yeah... we were just discussing this issue this morning. What about creating a helper function within the try block that gets the default values? That way they still are only defined in one place

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Reviewed 1 of 8 files at r1, 2 of 8 files at r4.
Review status: 6 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[Tests/Unit/Stubs/SQLServerStub.psm1, line  \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzYx6C0I6AaXlBLu2n:-KPEw1_bn7nF36MBY4dr:b-vydxgq) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/Stubs/SQLServerStub.psm1#L)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary>
> Shall I keep this file in?
</details>
Yeah, I think it's fine to keep it in

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 41 \[r4\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KP-QyMC8AO4S3OPGno7:-KPEzB0_SapVrzDKwAyv:b-77a7fu) ([raw file](https://github.com/powershell/xsqlserver/blob/4b6c96f3fa65059742997cdfa59b2edfe96f569d/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L41)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Yeah... we were just discussing this issue this morning. What about creating a helper function within the try block that gets the default values? That way they still are only defined in one place
</details>
Here's a [link](https://github.com/PowerShell/DscResources/issues/175) to the current discussion on a similar topic 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

aultt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 6 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-:-KPITg-NwzTzwLT_nNkh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Reviewed 2 of 8 files at r4, 2 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  I'm going to merge this for now, since all of the functionality looks good. We can work on the test formatting later once we have a solid workaround.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-:-KPKWdnCnwghwfRhLyRi:b-inuuky)*
<!-- Sent from Reviewable.io -->

mbreakey3 commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1, line 41 \[r4\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KP-QyMC8AO4S3OPGno7:-KPKWqTK71QQgWswqzz3:bb6ywmx) ([raw file](https://github.com/powershell/xsqlserver/blob/4b6c96f3fa65059742997cdfa59b2edfe96f569d/Tests/Unit/xSQLServerAvailabilityGroupListener.Tests.ps1#L41)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Here's a [link](https://github.com/PowerShell/DscResources/issues/175) to the current discussion on a similar topic 
</details>
LGTM

---

*[Tests/Unit/Stubs/SQLServerStub.psm1, line  \[r3\]](https://reviewable.io:443/reviews/powershell/xsqlserver/98#-KOzYx6C0I6AaXlBLu2n:-KPKWsEE8BLInDIXPH3h:b-85i64v) ([raw file](https://github.com/powershell/xsqlserver/blob/58869ddbd05670d6c6caecf68f99860952d646f0/Tests/Unit/Stubs/SQLServerStub.psm1#L)):*
<details><summary><i>Previously, mbreakey3 (Mariah) wrote…</i></summary>
> Yeah, I think it's fine to keep it in
</details>
LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsqlserver/98)*
<!-- Sent from Reviewable.io -->
